### PR TITLE
View Changelogs from Details Page

### PR DESCRIPTION
**Pull request by jbarlow96**
_**Purpose of this PR**_
As a user, I want to be able to view change logs from the details page so that I can have better context on what I'm looking at.

While the revision button is not toggle on, it will show all change logs of the released copy (so no change logs after last updated date will appear). If revision button is toggled on, it will show all change logs, even beyond the last updated date of the released learning object.

Other PR's involved with this - https://github.com/Cyber4All/CLARK-Gateway/pull/90 & https://github.com/Cyber4All/learning-object-service/pull/406

**While Signed In**
image

**While Signed Out**
image

**No Change Logs**
image

**Change Logs for Released Copy**
image

**Change Logs with Revision Button Toggled**
image

**NOTE: If this PR merges before Nick Winner's PR involving changing the shared module, then he needs to update the (click) event to (activate)**

### CODE REVIEW

**Comment by nrwinner** [comment]
@jbarlow96 Does this PR include the ability to filter the change logs by date? If the released object has a revision in progress, we only want to show change logs from up-to and before the object is released.

File: src/app/cube/details/splash/details-splash.component.html [hunk 1]
```diff
@@ -6,7 +6,9 @@
         <span tabindex="0">Created by </span>
         <a tabindex="0" [routerLink]="['/users', learningObject.author.username]" class="link" id="author-link" attr.aria-label="Clickable link {{ learningObject.author.name }}">{{learningObject.author.name | titlecase}}</a> - 
         <a tabindex="-0" [routerLink]="['/organization', learningObject.author.organization]" class="link"  attr.aria-label="Clickable link {{ learningObject.author.organization }}">{{learningObject.author.organization | titlecase}}</a>
-        <span tabindex="0">, updated {{ learningObject.date | date:'shortDate' }}</span>
+        <div class="updated" tabindex="0">Updated {{ learningObject.date | date:'shortDate' }}
+          <a *ngIf="canViewChangelogs" id="view-changelog" (click)="openViewAllChangelogsModal()" aria-label="Clickable link view the Changelog" tabindex="0">View the changelog</a>
```
Comment (pzalep1) [inline src/app/cube/details/splash/details-splash.component.html]: `<a *ngIf="canViewChangelogs" id="view-changelog" (click)="openViewAllChangelogsModal()" attr.aria-label="View the changelog for {{ learningObject.name }}" tabindex="0">View the changelog</a>`

File: src/app/cube/details/details.component.html [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <ng-container *ngIf="!loading.length; else loadingTemplate">
   <div *ngIf="learningObject && !errorStatus && revisedLearningObject" class="details-wrapper" [ngClass]="learningObject.length">
-    <cube-details-splash [canRate]="loggedin && auth.user.emailVerified && !isOwnObject" [ratings]="ratings" [averageRating]="averageRatingValue" [learningObject]="learningObject" (showNewRating)="showAddRating = $event;"></cube-details-splash>
+    <cube-details-splash [canRate]="loggedin && auth.user.emailVerified && !isOwnObject" [canViewChangelogs]="loggedin && auth.user.emailVerified" [ratings]="ratings" [averageRating]="averageRatingValue" [learningObject]="learningObject" (showNewRating)="showAddRating = $event;" (openChangelogModal)="openViewAllChangelogsModal()"></cube-details-splash>
```
Comment (nvisal1) [inline src/app/cube/details/details.component.html]: Is email verification required to view changelogs? If so, the backend business logic will need to be updated. Does anyone have context on this? @Cyber4All/front-end @DonSeannelly

File: src/app/cube/details/details.component.ts [hunk 3]
```diff
@@ -335,6 +352,35 @@ export class DetailsComponent implements OnInit, OnDestroy {
...
   }
 
+  /**
+   * Opens the Change Log modal for a specified learning object and fetches its change logs
+   */
+  async openViewAllChangelogsModal() {
+    if (!this.openChangelogModal ) {
+      this.openChangelogModal = true;
+      this.loadingChangelogs = true;
+      this.changelogLearningObject = this.learningObject;
```
Comment (nvisal1) [inline src/app/cube/details/details.component.ts]: This assignment is unnecessary. We can just use `this.learningObject` here.

**Comment by jbarlow96** [comment]
> @jbarlow96 Does this PR include the ability to filter the change logs by date? If the released object has a revision in progress, we only want to show change logs from up-to and before the object is released.

It does not because I totally forgot about that... I will work on adding that now.

File: src/app/cube/details/splash/details-splash.component.ts [hunk 4]
```diff
@@ -31,6 +33,13 @@ export class DetailsSplashComponent implements OnInit {
     this.showNewRating.emit(true);
   }
 
+  /**
+   * Emits an event to parent component to signal the change logs modal to open
+   */
+  openViewAllChangelogsModal() {
+    this.openChangelogModal.emit(true);
```
Comment (nvisal1) [inline src/app/cube/details/splash/details-splash.component.ts]: `true` is not being used by the parent component. You can just use `.emit()`.

File: src/app/cube/details/details.component.ts [hunk 5]
```diff
@@ -335,6 +352,35 @@ export class DetailsComponent implements OnInit, OnDestroy {
...
+      try {
+        this.changelogs = await this.changelogService.fetchAllChangelogs({
```
Comment (nvisal1) [inline src/app/cube/details/details.component.ts]: This function needs to be extended to take revisions into account. The Learning Object service endpoint accepts a `minusRevision` query parameter. The `fetchAllChangelogs` function will need that parameter to know how many changelogs to return to the client.

- If a user is viewing changelogs for a released copy of a Learning Object on the details page, `minusRevision` should be set to true. This tells the service to only return changelogs that are relevant to the released copy. 

- If a user is viewing changelogs for a revision (working copy) of a Learning Object on the details page, `minusRevision` should be set to false. This tells the service to return all changelogs that are relevant to the released copy and the working copy of the Learning Object.

**Comment by nvisal1** [comment]
> > @jbarlow96 Does this PR include the ability to filter the change logs by date? If the released object has a revision in progress, we only want to show change logs from up-to and before the object is released.
> 
> It does not because I totally forgot about that... I will work on adding that now.

@jbarlow96  @nrwinner Filtering by date has been deprecated in favor of the `minusRevision` query parameter. Refer to the comment below. 👍

File: src/app/cube/details/details.component.html (hunk 2 above)
Comment (jbarlow96) [reply src/app/cube/details/details.component.html]: I guess it wouldn't need to be in the sense that the user isn't leaving anything on the site (such as a rating), and is just viewing the history of a learning object. But I am open to hear other opinions on this.

Comment (nvisal1) [reply src/app/cube/details/details.component.html]: Email verification is required for Learning Object ratings to prevent potential bot spam.

**Comment by jbarlow96** [comment]
meme5

**Pull request merged**